sdk/iot/pnp/inc/az_iot_pnp_client.h

```diff
@@ -0,0 +1,334 @@
+
+/*
+ *
+ * PnP Command APIs
+ *
+ */
+
+/**
+ * @brief Azure IoT PnP command subscription options.
+ *
+ */
+typedef struct az_iot_pnp_client_command_subscription_options
+{
+  struct
+  {
+    int _not_used;
+  } _internal;
+} az_iot_pnp_client_command_subscription_options;
+
+/**
+ * @brief Gets the default Azure IoT PnP command subscription options.
+ * @details Call this to obtain an initialized #az_iot_pnp_client_command_subscription_options
+ * structure that can be afterwards modified and passed to
+ * #az_iot_pnp_client_commands_subscribe_topic_filter_get.
+ *
+ * @return #az_iot_pnp_client_command_subscription_options.
+ */
+AZ_NODISCARD az_iot_pnp_client_command_subscription_options az_iot_pnp_client_command_subscription_options_default();
+
+/**
+ * @brief Gets the MQTT topic filter to subscribe to PnP commands.
+ *
+ * @param[in] client The #az_iot_pnp_client to use for this call.
+ * @param[in] options A reference to an #az_iot_pnp_client_command_subscription_options structure.
+ * Can be NULL.
+ * @param[in] mqtt_topic_filter An empty #az_span with sufficient capacity to hold the MQTT topic
+ *                              filter.
+ * @param[out] out_mqtt_topic_filter The output #az_span containing the MQTT topic filter.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_pnp_client_commands_subscribe_topic_filter_get(
+    az_iot_pnp_client const* client,
+    az_iot_pnp_client_command_subscription_options const* options,
+    az_span mqtt_topic_filter,
+    az_span* out_mqtt_topic_filter);
+
+/**
+ * @brief A command request received from the PnP service.
+ *
+ */
+typedef struct az_iot_pnp_client_command_request
+{
+  az_span request_id; /**< The request id.
+                       * @note The application must match the method request and method response. */
+  az_span component; /**< The component the command is targeted towards. */
+  az_span command_name; /**< The name of the command to invoke */
+  az_span payload; /**< The body of the request for the command, stripped of any envelope data from
+                        MQTT payload */
+} az_iot_pnp_client_command_request;
+
+/**
+ * @brief Attempts to parse a received message's topic.
+ *
+ * @param[in] client The #az_iot_pnp_client to use for this call.
+ * @param[in] received_topic An #az_span containing the received topic.
+ * @param[in] received_payload An #az_span containing the received payload.
+ * @param[out] out_request If the message is a command request, this will contain the
+ *                         #az_iot_pnp_client_command_request.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_pnp_client_command_received_topic_parse(
+    az_iot_pnp_client const* client,
+    az_span received_topic,
+    az_span received_payload,
+    az_iot_pnp_client_command_request* out_command_request);
+
+/**
+ * @brief Azure IoT PnP command response options.
+ *
+ */
+typedef struct az_iot_pnp_client_command_response_options
+{
+  struct
+  {
+    int _not_used;
```

Comment:
I see these left with every operation. I think they make the public API a bit harder to understand/read. Could we also have versions of these APIs without the parameter that must be `NULL` today every single call? 
We could maybe postpone adding these options until we really need them?

Reply:
If you have no options today, how about this:
1. remove these *_options structs from the public API
2. remove the function that returns the default *_options struct instances.
3. modify the functions that takes these structs to take a void* called 'reserved' and document that NULL must be passed.

This will prevent customers from creating these things and calling functions that do nothing and force them to use the simpler code. Later, you can introduce these things publicly when you're ready.

Reply:
OK - agree void* is cleaner, will fix.  @JeffreyRichter should this guidance be put into the general API guidelines someplace so others don't re-invent a different pattern?

Reply:
One final question - I want to explicitly document that `reserved` must be NULL and put a pre-condition check on it in the implementation.  This way we can be sure all callers are setting it NULL if we ever allow non-NULL.  Allowing this to be anything would run risk of breaking anything for whatever reason was passing stuff in here.

Any concerns?